### PR TITLE
Base support for WASI in wasmer-js

### PR DESCRIPTION
**Pull request by syrusakbary**
This PR adds support for WASI targeting Javascript

### CODE REVIEW

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try

File: lib/vfs/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+edition = "2018"
+
+[dependencies]
+libc = { version = "^0.2", default-features = false, optional = true }
+thiserror = "1"
+tracing = { version = "0.1" }
+typetag = { version = "0.1", optional = true }
+serde = { version = "1.0", default-features = false, features = ["derive"], optional = true }
+
+[features]
```
Comment (Hywan) [inline lib/vfs/Cargo.toml:16]:
```suggestion
[features]
default = ["host_fs"]
```

Let's define `host_fs` as the default file system, which is already the case and the expected behavior.

File: lib/vfs/src/host_fs.rs [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+        fs::remove_dir(path).map_err(Into::into)
+    }
+    fn rename(&self, from: &Path, to: &Path) -> Result<(), FsError> {
+        fs::rename(from, to).map_err(Into::into)
+    }
+
+    fn remove_file(&self, path: &Path) -> Result<(), FsError> {
+        fs::remove_file(path).map_err(Into::into)
+    }
+    fn new_open_options(&self) -> OpenOptions {
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:43]:
```suggestion
    }

    fn new_open_options(&self) -> OpenOptions {
```

File: lib/vfs/src/host_fs.rs [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+    }
+
+    fn metadata(&self, path: &Path) -> Result<Metadata, FsError> {
+        fs::metadata(path)
+            .and_then(fs_metadata_to_metadata)
+            .map_err(Into::into)
+    }
+}
+
+fn fs_metadata_to_metadata(metadata: fs::Metadata) -> Result<Metadata, std::io::Error> {
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:54]: Can't we replace this by a `TryFrom` implementation instead? It would be more elegant.

File: lib/vfs/src/host_fs.rs [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+    fn open(
+        &mut self,
+        path: &Path,
+        conf: &OpenOptionsConfig,
+    ) -> Result<Box<dyn VirtualFile>, FsError> {
+        // TODO: handle create implying write, etc.
+        let read = conf.read();
+        let write = conf.write();
+        let append = conf.append();
+        let mut oo = std::fs::OpenOptions::new();
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:116]:
```suggestion
        let mut oo = fs::OpenOptions::new();
```

File: lib/vfs/src/host_fs.rs [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+    }
+
+    #[cfg(unix)]
+    fn get_raw_fd(&self) -> Option<i32> {
+        use std::os::unix::io::AsRawFd;
+        Some(self.inner.as_raw_fd())
+    }
+    #[cfg(not(unix))]
+    fn get_raw_fd(&self) -> Option<i32> {
+        unimplemented!(
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:351]: I believe we may be able to use [`AaRawHandle`](https://doc.rust-lang.org/std/os/windows/io/trait.AsRawHandle.html) and [`RawHandle`](https://doc.rust-lang.org/std/os/windows/io/type.RawHandle.html) here. It's the size of a pointer (`type RawHandle = HANDLE;` and `type HANDLE = *mut c_void`), so we should update `i32` that represents a file descriptor.

Ideally, we should come with a type to represent a file descriptor:

```rust
type FileDescriptor = usize;

#[cfg(unix)]
impl From<RawFd> for FileDescriptor { … }

#[cfg(windows)]
impl From<RawHandle> for FileDescriptor { … }
```

File: lib/vfs/src/host_fs.rs [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+
+        host_file_bytes_available(host_fd)
+    }
+
+    #[cfg(unix)]
+    fn get_raw_fd(&self) -> Option<i32> {
+        use std::os::unix::io::AsRawFd;
+        Some(self.inner.as_raw_fd())
+    }
+    #[cfg(not(unix))]
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:349]:
```suggestion
    #[cfg(windows)]
```

File: lib/vfs/src/host_fs.rs [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+        {
+            use std::os::unix::fs::FileTypeExt;
+            (
+                filetype.is_char_device(),
+                filetype.is_block_device(),
+                filetype.is_socket(),
+                filetype.is_fifo(),
+            )
+        }
+        #[cfg(not(unix))]
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:68]:
```suggestion
        #[cfg(windows)]
```

File: lib/vfs/src/host_fs.rs [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+        // success
+        0 => Ok(bytes_found.try_into().unwrap_or(0)),
+        libc::EBADF => Err(FsError::InvalidFd),
+        libc::EFAULT => Err(FsError::InvalidData),
+        libc::EINVAL => Err(FsError::InvalidInput),
+        _ => Err(FsError::IOError),
+    }
+}
+
+#[cfg(not(unix))]
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:372]:
```suggestion
#[cfg(windows)]
```

File: lib/vfs/src/host_fs.rs [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+        let host_fd = self.get_raw_fd().unwrap();
+
+        host_file_bytes_available(host_fd)
+    }
+    #[cfg(unix)]
+    fn get_raw_fd(&self) -> Option<i32> {
+        use std::os::unix::io::AsRawFd;
+        Some(io::stdout().as_raw_fd())
+    }
+    #[cfg(not(unix))]
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:460]:
```suggestion
    #[cfg(windows)]
```

File: lib/vfs/src/host_fs.rs [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+        let host_fd = self.get_raw_fd().unwrap();
+
+        host_file_bytes_available(host_fd)
+    }
+    #[cfg(unix)]
+    fn get_raw_fd(&self) -> Option<i32> {
+        use std::os::unix::io::AsRawFd;
+        Some(io::stderr().as_raw_fd())
+    }
+    #[cfg(not(unix))]
```
Comment (Hywan) [inline lib/vfs/src/host_fs.rs:551]:
```suggestion
    #[cfg(windows)]
```

and so on

File: lib/vfs/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+use std::io::{self, Read, Seek, Write};
+use std::path::{Path, PathBuf};
+use thiserror::Error;
+use tracing::debug;
+
+#[cfg(feature = "host_fs")]
+pub mod host_fs;
+//pub mod vfs_fs;
+#[cfg(feature = "mem_fs")]
+pub mod mem_fs;
```
Comment (Hywan) [inline lib/vfs/src/lib.rs:13]: I think both features are mutually exclusive.

We should add something like this:

```rust
#[cfg(all(feature = "host_fs", feature = "mem_fs"))]
compile_error!("…");
```

Thoughts?

File: lib/vfs/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+use std::ffi::OsString;
+use std::fmt;
+use std::io::{self, Read, Seek, Write};
+use std::path::{Path, PathBuf};
+use thiserror::Error;
+use tracing::debug;
+
+#[cfg(feature = "host_fs")]
+pub mod host_fs;
+//pub mod vfs_fs;
```
Comment (Hywan) [inline lib/vfs/src/lib.rs:11]:
```suggestion
```

File: lib/vfs/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+
+    /// the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn created_time(&self) -> u64;
+
+    /// the size of the file in bytes
+    fn size(&self) -> u64;
+
+    /// Change the size of the file, if the `new_size` is greater than the current size
+    /// the extra bytes will be allocated and zeroed
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError>;
```
Comment (Hywan) [inline lib/vfs/src/lib.rs:149]: Usually, this method is named `truncate`. I think it's convey the intend much better. Thought?

File: lib/vfs/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+            io::ErrorKind::Other => FsError::IOError,
+            // if the following triggers, a new error type was added to this non-exhaustive enum
+            _ => FsError::UnknownError,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ReadDir {
+    // TODO: to do this properly we need some kind of callback to the core FS abstraction
```
Comment (Hywan) [inline lib/vfs/src/lib.rs:312]: Need to be addressed, right?

File: lib/vfs/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+impl ReadDir {
+    pub fn new(data: Vec<DirEntry>) -> Self {
+        Self { data, index: 0 }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DirEntry {
+    pub path: PathBuf,
+    // weird hack, to fix this we probably need an internal trait object or callbacks or something
```
Comment (Hywan) [inline lib/vfs/src/lib.rs:326]: Why is it a weird hack?

File: lib/vfs/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,443 @@
...
+    pub metadata: Result<Metadata, FsError>,
+}
+
+impl DirEntry {
+    pub fn path(&self) -> PathBuf {
+        self.path.clone()
```
Comment (Hywan) [inline lib/vfs/src/lib.rs:332]: Why cloning?

File: lib/vfs/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,443 @@
...
+    }
+
+    pub fn metadata(&self) -> Result<Metadata, FsError> {
+        self.metadata.clone()
```
Comment (Hywan) [inline lib/vfs/src/lib.rs:336]: Why cloning?

File: lib/vfs/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,443 @@
...
+    pub fn file_name(&self) -> OsString {
+        self.path
+            .file_name()
+            .unwrap_or_else(|| self.path.as_os_str())
+            .to_owned()
+    }
+}
+
+#[derive(Clone, Debug)]
+// TODO: review this, proper solution would probably use a trait object internally
```
Comment (Hywan) [inline lib/vfs/src/lib.rs:353]: I don't understand what's wrong with this code. Can you elaborate please?

Comment (Hywan) [reply lib/vfs/src/lib.rs:353]: Oh I see, maybe we should implement `Metadata` as a trait, and not as a struct. And I think it's a good idea indeed!

File: lib/vfs/src/vfs_fs.rs [hunk 19]
```diff
@@ -0,0 +1,183 @@
+use crate::*;
+use std::io::{Read, Seek, Write};
+use std::path::Path;
+use std::sync::Arc;
+
+pub use crate::host_fs::{Stderr, Stdin, Stdout};
+
+#[derive(Clone)]
+pub struct VfsFileSystem {
```
Comment (Hywan) [inline lib/vfs/src/vfs_fs.rs:9]: What's the difference between a virtual file system file system (I'm just expandind the name here :-p), and an in-memory file system?

File: lib/wasi-experimental-io-devices/Cargo.toml [hunk 20]
```diff
@@ -14,7 +14,7 @@ edition = "2018"
 maintenance = { status = "experimental" }
 
 [dependencies]
-wasmer-wasi = { version = "2.0.0", path = "../wasi" }
+wasmer-wasi = { version = "2.0.0", path = "../wasi", default-features=false }
```
Comment (Hywan) [inline lib/wasi-experimental-io-devices/Cargo.toml:17]: Why don't we want the default features?

File: lib/wasi/Cargo.toml [hunk 21]
```diff
@@ -10,22 +10,52 @@ license = "MIT"
...
 logging = ["tracing/log"]
-disable-all-logging = ["tracing/release_max_level_off", "tracing/max_level_off"]
\ No newline at end of file
+disable-all-logging = [
+    "tracing/release_max_level_off",
+    "tracing/max_level_off"
+]
+host_fs = ["wasmer-vfs/host_fs"]
+mem_fs = ["wasmer-vfs/mem_fs"]
+js = [
```
Comment (Hywan) [inline lib/wasi/Cargo.toml:47]: Just like we did with `wasmer` (`lib/api/`), I think it's good to differentiate between `sys` and `js`. Thought?

File: lib/wasi/Cargo.toml [hunk 22]
```diff
@@ -10,22 +10,52 @@ license = "MIT"
...
+]
+test-js = [
+    "js",
+    "wasmer/js-default",
+    "wasmer/wat"
+]
+
+[dev-dependencies]
+wasm-bindgen = "0.2"
+wasm-bindgen-test = "0.3.0"
```
Comment (Hywan) [inline lib/wasi/Cargo.toml:60]:
```suggestion
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-bindgen = "0.2"

[target.'cfg(target_arch = "wasm32")'.dev-dependencies]
wasm-bindgen-test = "0.3.0"
```

`wasm-bindgen-test` is a `dev-dependencies`. And both are for `wasm32` only.

File: lib/wasi/src/macros.rs [hunk 23]
```diff
@@ -31,6 +31,6 @@ macro_rules! wasi_try {
 /// corrupted.
 macro_rules! get_input_str {
     ($memory:expr, $data:expr, $len:expr) => {{
-        wasi_try!($data.get_utf8_str($memory, $len), __WASI_EINVAL)
+        wasi_try!($data.get_utf8_string($memory, $len), __WASI_EINVAL)
```
Comment (Hywan) [inline lib/wasi/src/macros.rs:34]: Why not keeping `get_utf8_str`? It returns a `Cow`, so you can easily transform it into a `String` if needed.

File: lib/wasi/src/ptr.rs [hunk 24]
```diff
@@ -82,7 +82,16 @@ impl<T: Copy + ValueType> WasmPtr<T, Array> {
...
+        self,
+        memory: &'a Memory,
+        str_len: u32,
+    ) -> Option<std::borrow::Cow<'a, str>> {
+        self.0.get_utf8_str(memory, str_len).map(Into::into)
+    }
+
+    #[inline(always)]
+    pub unsafe fn get_utf8_string(self, memory: &Memory, str_len: u32) -> Option<String> {
+        self.0.get_utf8_string(memory, str_len)
```
Comment (Hywan) [inline lib/wasi/src/ptr.rs:95]: Not tested, but I reckon it can be written as follows:

```suggestion
        self.get_utf8_str(memory, str_len)?.to_owned()
```

What do you think?

File: lib/wasi/src/state/builder.rs [hunk 25]
```diff
@@ -37,15 +38,18 @@ pub struct WasiStateBuilder {
...
-    stdin_override: Option<Box<dyn WasiFile>>,
+    stdout_override: Option<Box<dyn VirtualFile>>,
+    stderr_override: Option<Box<dyn VirtualFile>>,
+    stdin_override: Option<Box<dyn VirtualFile>>,
+    fs_override: Option<Box<dyn wasmer_vfs::FileSystem>>,
 }
 
 impl std::fmt::Debug for WasiStateBuilder {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // TODO: update this when stable
```
Comment (Hywan) [inline lib/wasi/src/state/builder.rs:52]: ⚠️ There is a todo here that must be addressed at the end of the review process.

File: lib/wasi/src/state/builder.rs [hunk 26]
```diff
@@ -76,7 +80,7 @@ pub enum WasiStateCreationError {
     #[error("wasi filesystem setup error: `{0}`")]
     WasiFsSetupError(String),
     #[error(transparent)]
-    WasiFsError(WasiFsError),
+    FsError(FsError),
```
Comment (Hywan) [inline lib/wasi/src/state/builder.rs:83]: The other variants are named `WasiFs*`, so let's keep `WasiFsError` here?

File: lib/wasi/src/state/builder.rs [hunk 27]
```diff
@@ -359,29 +385,35 @@ impl WasiStateBuilder {
             }
         }
 
-        // self.preopens are checked in [`PreopenDirBuilder::build`]
+        // using `take` here is a bit of a code smell
```
Comment (Hywan) [inline lib/wasi/src/state/builder.rs:388]: Yeah, totally agree. If the `build` method was taking ownership of `self` rather than a mutable reference, that would not be an issue because `self` wouldn't be any longer usable. But in this case, it changes the behavior of the object. I don't like it at all :-). It must be changed!

File: lib/wasi/src/state/mod.rs [hunk 28]
```diff
@@ -163,12 +168,111 @@ pub struct WasiFs {
...
+        } else if #[cfg(feature = "mem_fs")] {
+            Box::new(wasmer_vfs::mem_fs::MemFileSystem::default())
+        } else {
+            Box::new(FallbackFileSystem::default())
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct FallbackFileSystem;
```
Comment (Hywan) [inline lib/wasi/src/state/mod.rs:189]: `FallbackFileSystem` is not required if at least `host_fs` _or_ `mem_fs` is required. One of them is required. But they are mutually exclusive.

I think it's the expected behavior here. WASI without a FS is… useless…, right?

File: lib/wasi/src/syscalls/mod.rs [hunk 29]
```diff
@@ -90,9 +96,10 @@ fn read_bytes<T: Read>(
         bytes_read += reader.read(&mut raw_bytes).map_err(|_| __WASI_EIO)? as u32;
         unsafe {
             memory
-                .view::<u8>()
+                .uint8view()
```
Comment (Hywan) [inline lib/wasi/src/syscalls/mod.rs:99]: Let's keep `view::<u8>` here, it's just simple.

**Review by Hywan — CHANGES_REQUESTED** [review]

**Comment by Hywan** [comment]
This PR adds the following:

* It introduces `wasmer-vfs` (from @MarkMcCaskey I suppose). Please, present this crate a little bit and provide examples :-).
* It introduces the `--verbose` option to `wasmer` CLI,
* It allows to transform `ImportObject` to a JavaScript object,
* It removes `serde` from `wasmer-c-api`, which wasn't exposed to the public API,
* It uses `wasmer-vfs` in `wasmer-wasi`,
* It introduces JavaScript in `wasmer-wasi`?

I didn't review `memfs` and `vfsfs` because I don't know the difference betweem them, and I suspect one of them are going to be removed :-).

It's also too much to review. Can we split this PR in 2 parts:

1. `wasmer-vfs` + VFS in WASI,
2. WASI in JavaScript.

please?

Can't we run all WASI tests with `wasm-bindgen-tests`? That would be terrible.

File: lib/vfs/Cargo.toml [hunk 30]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "wasmer-vfs"
+version = "2.0.0"
+description = "Wasmer Virtual FileSystem"
```
[the remark this answers is not in the document]
Comment (syrusakbary) [reply lib/vfs/Cargo.toml:4]: I want the VirtualFS implementation to be separated from WASI, and usable standalone.
I think is better without the WASI mention

File: lib/wasi/src/state/builder.rs (hunk 26 above)
Comment (syrusakbary) [reply lib/wasi/src/state/builder.rs:83]: In this case I believe it will not be a WasiFsError but a FileSystem error. I agree we need some consistency on the error names though

File: lib/wasi/src/state/mod.rs (hunk 28 above)
Comment (syrusakbary) [reply lib/wasi/src/state/mod.rs:189]: Not at all.. some commands (like cowsay) operate without a filesystem... they just do compute :)

File: lib/vfs/src/lib.rs (hunk 12 above)
Comment (Hywan) [reply lib/vfs/src/lib.rs:11]: See https://github.com/wasmerio/wasmer/pull/2522 to solve this comment.

File: lib/vfs/src/lib.rs (hunk 11 above)
Comment (Hywan) [reply lib/vfs/src/lib.rs:13]: See https://github.com/wasmerio/wasmer/pull/2523 to solve this comment.

File: lib/vfs/Cargo.toml (hunk 1 above)
Comment (Hywan) [reply lib/vfs/Cargo.toml:16]: See https://github.com/wasmerio/wasmer/pull/2523 to solve this comment.

File: lib/vfs/Cargo.toml (hunk 30 above)
[the remark this answers is not in the document]
Comment (Hywan) [reply lib/vfs/Cargo.toml:4]: OK.

File: lib/cli/Cargo.toml [hunk 31]
```diff
@@ -35,11 +35,12 @@ wasmer-engine-universal = { version = "2.0.0", path = "../engine-universal", opt
 wasmer-engine-dylib = { version = "2.0.0", path = "../engine-dylib", optional = true }
 wasmer-engine-staticlib = { version = "2.0.0", path = "../engine-staticlib", optional = true }
 wasmer-vm = { version = "2.0.0", path = "../vm" }
-wasmer-wasi = { version = "2.0.0", path = "../wasi", default-features = false, optional = true }
+wasmer-wasi = { version = "2.0.0", path = "../wasi", default-features = false, features = ["host_fs"], optional = true }
```
[the remark this answers is not in the document]
Comment (Hywan) [reply lib/cli/Cargo.toml:38]: OK, so previously we were using the default features of `wasmer-wasi`. Now, `wasmer-wasi` has `host_fs` as the default feature, so that the previous behavior is kept, which is the correct way to do.

So… we don't need to update `wasmer-c-api` and `wasmer-cli` here. It's been fixed in https://github.com/wasmerio/wasmer/pull/2523.

File: lib/c-api/Cargo.toml [hunk 32]
```diff
@@ -32,7 +32,7 @@ wasmer-engine-universal = { version = "2.0.0", path = "../engine-universal", opt
 wasmer-engine-dylib = { version = "2.0.0", path = "../engine-dylib", optional = true }
 wasmer-engine-staticlib = { version = "2.0.0", path = "../engine-staticlib", optional = true }
 wasmer-middlewares = { version = "2.0.0", path = "../middlewares", optional = true }
-wasmer-wasi = { version = "2.0.0", path = "../wasi", optional = true }
+wasmer-wasi = { version = "2.0.0", path = "../wasi", default-features = false, features = ["host_fs"], optional = true }
```
[the remark this answers is not in the document]
Comment (Hywan) [reply lib/c-api/Cargo.toml:35]: See https://github.com/wasmerio/wasmer/pull/2491#discussion_r689562448, which is similiar to this situation.

tl;dr: Fixed in https://github.com/wasmerio/wasmer/pull/2523.

File: lib/vfs/src/host_fs.rs (hunk 6 above)
Comment (Hywan) [reply lib/vfs/src/host_fs.rs:350]: See https://github.com/wasmerio/wasmer/pull/2524 to fix it. It explains why `not(unix)` is not equivalent to `windows` (tl;dr it's because there is the `wasm` target family, which is new).

File: lib/vfs/src/host_fs.rs (hunk 7 above)
Comment (Hywan) [reply lib/vfs/src/host_fs.rs:69]: See https://github.com/wasmerio/wasmer/pull/2491#discussion_r689570334. tl;dr it's fixed by https://github.com/wasmerio/wasmer/pull/2524.

File: lib/vfs/src/host_fs.rs (hunk 8 above)
Comment (Hywan) [reply lib/vfs/src/host_fs.rs:373]: See https://github.com/wasmerio/wasmer/pull/2491#discussion_r689570334. tl;dr it's fixed by https://github.com/wasmerio/wasmer/pull/2524.

File: lib/vfs/src/host_fs.rs (hunk 9 above)
Comment (Hywan) [reply lib/vfs/src/host_fs.rs:461]: See https://github.com/wasmerio/wasmer/pull/2491#discussion_r689570334. tl;dr it's fixed by https://github.com/wasmerio/wasmer/pull/2524.

File: lib/vfs/src/host_fs.rs (hunk 10 above)
Comment (Hywan) [reply lib/vfs/src/host_fs.rs:552]: See https://github.com/wasmerio/wasmer/pull/2491#discussion_r689570334. tl;dr it's fixed by https://github.com/wasmerio/wasmer/pull/2524.

File: lib/cli/src/commands/run.rs [hunk 33]
```diff
@@ -64,6 +64,9 @@ pub struct Run {
     #[structopt(long = "debug", short = "d")]
     debug: bool,
 
+    #[structopt(short, long, parse(from_occurrences))]
```
[the remark this answers is not in the document]
Comment (Hywan) [reply lib/cli/src/commands/run.rs:67]: I've not noticed the `short` option already. Marking this comment as resolved.

Testing to confirm:

```shell
$ ./target/release/wasmer run -h | rg verbose
    -v, --verbose                         
```

The `-v` option is the short version of `--verbose`.

File: lib/vfs/src/host_fs.rs (hunk 3 above)
Comment (Hywan) [reply lib/vfs/src/host_fs.rs:55]: Because of the orphan Rust rule, it's better to implement `TryInto` rather than `TryFrom`.

Comment (Hywan) [reply lib/vfs/src/host_fs.rs:55]: Solved in https://github.com/wasmerio/wasmer/pull/2525.

File: lib/vfs/src/host_fs.rs (hunk 5 above)
Comment (Hywan) [reply lib/vfs/src/host_fs.rs:352]: Draft here, https://github.com/wasmerio/wasmer/pull/2528

**Comment by Hywan** [comment]
`wasmer-vfs` with serde enabled doesn't run.

**Comment by Hywan** [comment]
bors try

**Comment by syrusakbary** [comment]
bors try-

**Comment by syrusakbary** [comment]
bors try

**Comment by Hywan** [comment]
bors try

File: lib/api/src/sys/externals/memory.rs [hunk 34]
```diff
@@ -226,6 +226,11 @@ impl Memory {
         unsafe { MemoryView::new(base as _, length as u32) }
     }
 
+    /// A uint8view
+    pub fn uint8view(&self) -> MemoryView<u8> {
+        self.view()
+    }
+
```
[the remark this answers is not in the document]
Comment (Hywan) [reply lib/api/src/sys/externals/memory.rs]: Method is removed.

File: lib/wasi/src/macros.rs (hunk 23 above)
Comment (syrusakbary) [reply lib/wasi/src/macros.rs:34]: I forgot why, we can try to use `get_utf8_str` but I believe I did this because the other didn't work in both js and sys

File: lib/wasi/src/state/mod.rs [hunk 35]
```diff
@@ -1428,16 +1572,18 @@ impl WasiState {
 
     /// Turn the WasiState into bytes
     pub fn freeze(&self) -> Option<Vec<u8>> {
-        bincode::serialize(self).ok()
+        todo!("temporarily disabled")
+        //bincode::serialize(self).ok()
```
Comment (Hywan) [inline lib/wasi/src/state/mod.rs:1576]: Why? cc @syrusakbary

File: lib/wasi/src/syscalls/mod.rs [hunk 36]
```diff
@@ -2512,10 +2498,15 @@ pub fn poll_oneoff(
...
+//     // RuntimeError::raise(Box::new(WasiError::Exit(code)));
+// }
 pub fn proc_exit(env: &WasiEnv, code: __wasi_exitcode_t) {
-    debug!("wasi::proc_exit, {}", code);
-    RuntimeError::raise(Box::new(WasiError::Exit(code)));
-    unreachable!();
+    // panic!("Exit: {}", code);
+    // Err(WasiError::Exit(code))
+    // RuntimeError::raise(Box::new(WasiError::Exit(code)));
 }
```
Comment (Hywan) [inline lib/wasi/src/syscalls/mod.rs:2510]: We need to clarify this.

**Comment by Hywan** [comment]
bors try

File: lib/wasi/src/syscalls/mod.rs (hunk 36 above)
Comment (Hywan) [reply lib/wasi/src/syscalls/mod.rs]: Fixed in https://github.com/wasmerio/wasmer/pull/2491/commits/650352ddeca4e130fcfe45312f4a9ccdd1cea841.

File: lib/wasi/src/state/mod.rs (hunk 35 above)
Comment (syrusakbary) [reply lib/wasi/src/state/mod.rs:1576]: We can revert and make bincode an optional dependency if `enable-serde` feature is set

File: lib/wasi/src/state/mod.rs [hunk 37]
```diff
@@ -1428,16 +1572,18 @@ impl WasiState {
...
     }
 
     /// Get a WasiState from bytes
-    pub fn unfreeze(bytes: &[u8]) -> Option<Self> {
-        bincode::deserialize(bytes).ok()
+    pub fn unfreeze(_bytes: &[u8]) -> Option<Self> {
+        todo!("temporarily disabled")
+        //bincode::deserialize(bytes).ok()
```
Comment (syrusakbary) [inline lib/wasi/src/state/mod.rs:1582]: We can revert and make bincode an optional dependency if enable-serde feature is set

File: lib/wasi/src/state/builder.rs (hunk 27 above)
Comment (syrusakbary) [reply lib/wasi/src/state/builder.rs:388]: Resolution: make `build` consume `self` (so we don't need to wrap in an `Arc` or other paths)

Comment (Hywan) [reply lib/wasi/src/state/builder.rs]: It's not possible because it will break existing code, like:

```
let mut builder = WasiStateBuilder::new();
builder.stdin(…);
builder.set_fs(…);
```

it won't compile anymore because `stdin` will consume `self`.

Except if we introduce a BC break here, it's not possible to solve this problem. I've decided to write what's happening in the documentation when calling `build` or `finalize` multiple times. I think it's the best trade-off for the moment.

See https://github.com/wasmerio/wasmer/pull/2491/commits/4eaa864ae8b08daa3dd97f97c633ad0e6a6b6583.

File: lib/wasi/src/state/mod.rs (hunk 37 above)
Comment (Hywan) [reply lib/wasi/src/state/mod.rs]: Partially solved in https://github.com/wasmerio/wasmer/pull/2491/commits/a12ec28029ea902f13795e5d573f49b82a670556.

`wasmer-wasi` with `enable-serde` doesn't compile for the moment, but it's not new. Let's address that in another PR.

**Review by Hywan — APPROVED** [review]

**Comment by Hywan** [comment]
I'm not fully satisfied with the state of this PR, but it's way better than it was initially. Let's fix the other pain points in other PR.

bors r+
